Add save/load buttons for the VTX Table

McGiverGim opened the pull request:
It adds a load/save button for the VTX Tables in the VTX tab that work with JSON format.

![image](https://user-images.githubusercontent.com/2673520/64105083-d03c6f00-cd75-11e9-8e40-05ce3fc36201.png)

A sample of the file generated is:
```json
{
    "description": "Betaflight VTX Config file",
    "version": "1.0",
    "vtx_table": {
        "bands": 5,
        "channels": 8,
        "bands_list": [
            {
                "number": 1,
                "name": "BOSCAM_A",
                "letter": "A",
                "is_factory_band": true,
                "frequencies": [
                    5865,
                    5845,
                    5825,
                    5805,
                    5785,
                    5765,
                    5745,
                    5725
                ]
            },
            {
                "number": 2,
                "name": "BOSCAM_B",
                "letter": "B",
                "is_factory_band": true,
                "frequencies": [
                    5733,
                    5752,
                    5771,
                    5790,
                    5809,
                    5828,
                    5847,
                    5866
                ]
            },
            {
                "number": 3,
                "name": "BOSCAM_E",
                "letter": "E",
                "is_factory_band": true,
                "frequencies": [
                    5705,
                    5685,
                    5665,
                    0,
                    5885,
                    5905,
                    0,
                    0
                ]
            },
            {
                "number": 4,
                "name": "FATSHARK",
                "letter": "F",
                "is_factory_band": true,
                "frequencies": [
                    5740,
                    5760,
                    5780,
                    5800,
                    5820,
                    5840,
                    5860,
                    5880
                ]
            },
            {
                "number": 5,
                "name": "RACEBAND",
                "letter": "R",
                "is_factory_band": true,
                "frequencies": [
                    5658,
                    5695,
                    5732,
                    5769,
                    5806,
                    5843,
                    5880,
                    5917
                ]
            }
        ],
        "powerlevels": 4,
        "powerlevels_list": [
            {
                "number": 1,
                "value": 0,
                "label": "25 "
            },
            {
                "number": 2,
                "value": 1,
                "label": "200"
            },
            {
                "number": 3,
                "value": 2,
                "label": "600"
            },
            {
                "number": 4,
                "value": 3,
                "label": "1.2"
            }
        ]
    }
}
```

The buttons work as other tabs:
- They save/load the content but only a message in the GUI log appear. Maybe is better to add a modal window with that info? We don't do it at other similar tabs.
- It saves the "live" tab content (not the saved in the firmware one) to the file. I thought this is most intuitive to the user.
- It loads the content to the tab, but it does not save it. Maybe we need to add a GUI message/modal window saying the user to verify/modify/save it.

McGiverGim commented:
Force pushed a PR to simplify the detection of write end, and removing some forgotten copy&paste :(

mikeller commented:
Thanks for this @McGiverGim.

I think in the JSON format, we can drop `bands`, `powerlevels`, and `number` in the individual entries - they are all implicit, and if we keep them that only means we need to implement a way to deal with inconsistencies.

After loading we can probably display a message on the tab, saying 'VTX settings loaded from <filename>, use 'Save' to write them to the flight controller.' or similar. This will give the user feedback about what has happened, and remind them that the settings need to be saved.

And yes, I agree that saving the data that is shown to the user is the correct thing to do.

McGiverGim commented:
Ok, I will remove the useless elements. The message to the user, are you saying to show it as a "note" in the tab?

McGiverGim commented:
Finally, I added some changes:
1. Refactor the code to simplify the way the content is loaded/saved from the JSON. In this way the maintainability is greater and we can add new fonts of VTX Tables easier. In the original way each font must use JQuery to complete the table, now it only needs to populate one object. The only drawback is that now we "reload" the html. I can revert it if needed.
2. Simplified the JSON produced, sample:
```json
{
    "description": "Betaflight VTX Config file",
    "version": "1.0",
    "vtx_table": {
        "bands_list": [
            {
                "name": "BOSCAM_A",
                "letter": "A",
                "is_factory_band": true,
                "frequencies": [
                    5865,
                    5845,
                    5825,
                    5805,
                    5785,
                    5765,
                    5745,
                    5725
                ]
            },
            {
                "name": "BOSCAM_B",
                "letter": "B",
                "is_factory_band": true,
                "frequencies": [
                    5733,
                    5752,
                    5771,
                    5790,
                    5809,
                    5828,
                    5847,
                    5866
                ]
            },
            {
                "name": "BOSCAM_E",
                "letter": "E",
                "is_factory_band": true,
                "frequencies": [
                    5705,
                    5685,
                    5665,
                    5645,
                    5885,
                    5905,
                    5925,
                    5945
                ]
            },
            {
                "name": "FATSHARK",
                "letter": "F",
                "is_factory_band": true,
                "frequencies": [
                    5740,
                    5760,
                    5780,
                    5800,
                    5820,
                    5840,
                    5860,
                    5880
                ]
            },
            {
                "name": "RACEBAND",
                "letter": "R",
                "is_factory_band": true,
                "frequencies": [
                    5658,
                    5695,
                    5732,
                    5769,
                    5806,
                    5843,
                    5880,
                    5917
                ]
            }
        ],
        "powerlevels_list": [
            {
                "value": 0,
                "label": "25 "
            },
            {
                "value": 1,
                "label": "50 "
            }
        ]
    }
}
```
3. Added a note to remember the user to verify and save the loaded values:
![image](https://user-images.githubusercontent.com/2673520/64158567-04c52f00-ce39-11e9-9885-6b0d0a9ff838.png)

Docteh commented:
Fonts?

McGiverGim commented:
Sorry! Updated my comment, font->source

Bad Spanish-English translation :)

mikeller commented:
This looks mostly good, good work @McGiverGim.

Some things I found when testing:
- adding a new band or power level, and then saving saves some values as `null`. It's probably better to pre-process the values and set them to 0, like we do before saving them to the VTX.

```
            {
                "name": "",
                "letter": "",
                "is_factory_band": false,
                "frequencies": [
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null
                ]
            }
```

```
            {
                "value": null,
                "label": ""
            }
```

- when saving, the suggested file name does not have a `.json` extension - this will make it hard for users to save and then load their settings, as the default filtering is for `.json`;
- maybe we should add a timestamp to the suggested file name, like we do in other places? This will make it easy for users to store multiple versions into different files (see https://github.com/betaflight/betaflight-configurator/blob/054ab5e02aa57646feff2901bbe597ea1e42fe06/src/js/tabs/cli.js#L153).

McGiverGim commented:
> adding a new band or power level, and then saving saves some values as null. It's probably better to pre-process the values and set them to 0, like we do before saving them to the VTX.

Fixed, I added some default values to some of the fields. It fixes some old glitches.

> when saving, the suggested file name does not have a .json extension - this will make it hard for users to save and then load their settings, as the default filtering is for .json;

In my computer it was adding the json extension, I don't know why in yours not. A difference windows vs linux? No matter, I have implemented the third comment.

> maybe we should add a timestamp to the suggested file name, like we do in other places? This will make it easy for users to store multiple versions into different files

Fixed. Test it to see if you like it. I'm not too sure about the final name.

Docteh commented:
`vtx_config`? Personally I feel that `vtxtable` might be better. 

Right now its `BTFL_vtx_config_20190903_141817.json` but `BTFL_vtxtable_20190903_141817.json`

Any thoughts on trying to `JSON.parse` the contents of the clipboard?

mikeller commented on locales/en/messages.json:
```diff
@@ -4784,6 +4784,10 @@
         "description": "Message to show when the VTX is not supported in the VTX tab"
     },
+    "vtxMessageVerifyTable": {
+        "message": "<span class=\"message-negative\">Attention:</span> The values of the VTX Table have been loaded but not saved, you must verify and modify the values to be sure that that they are valid and legal at your country and then press the $t(vtxButtonSave.message) button.",
```
I would probably go for this here:

The values of the VTX Table have been loaded, but not yet stored on the flight controller. You must verify and modify the values to be sure that that they are valid and legal in your country and then press the Save button to store them on the flight controller.

Reason: 'Save' has two meanings in this context.

mikeller commented:
This is working fine now. I agree with @Docteh `vtxtable` or `vtx_table` is probably more appropriate as this is what we are calling the UI element. But the name is just a suggestion anyway, so it's not super important.

@Docteh: Import from clipboard is another way we can look into supporting, as well as adding support for more file formats ('CLI `diff` snippet' comes to mind, or 'compressed with checksum', so that an entire VTX table can be stored in a QR code and printed on VTX packaging by manufacturers. I think the important step that we are taking here is to support one basic format, which will allow manufacturers / communities / users to easily publish and share VTX tables, which will be crucial for this to be as easy as possible to use for users.

McGiverGim commented:
Fixed as per reviews, modified the text and chaged the filename format. Squashed too.

And yes, we can add easily the clipboard part, is only read it and call the same method that the load file does, but I think is better to do it in other PR.
